src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -5322,6 +5322,286 @@ class StubGenerator: public StubCodeGenerator {
+        255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u,
+        255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u,
+    };
+
+    __ align(CodeEntryAlignment);
+    StubCodeMark mark(this, "StubRoutines", "decodeBlock");
+    address start = __ pc();
+    __ enter();
+
+    Register src    = c_rarg0;
+    Register soff   = c_rarg1;
+    Register send   = c_rarg2;
+    Register dst    = c_rarg3;
+    Register doff   = c_rarg4;
+    Register isURL  = c_rarg5;
+    Register isMIME = c_rarg6;
+
+    Register codec     = c_rarg7;
+    Register dstBackup = x31;
+    Register length    = x28;     // t3, total length of src data in bytes
+
+    Label ProcessData, Exit;
+    Label ProcessScalar, ProcessVector;
+
+    // passed in length (send - soff) is guaranteed to be > 4,
+    // and in this intrinsic we only processe data of length in multiple of 4,
+    // it's not guaranteed to be multiple of 4 by java level, so do it explicitly
+    __ sub(length, send, soff);
+    __ andi(length, length, -4);
+    // real src/dst to process data
+    __ add(src, src, soff);
+    __ add(dst, dst, doff);
+    // backup of dst, used to calculate the return value at exit
+    __ mv(dstBackup, dst);
+
+    // load the codec base address
+    __ la(codec, ExternalAddress((address) fromBase64));
+    __ beqz(isURL, ProcessData);
+    __ la(codec, ExternalAddress((address) fromBase64URL));
+    __ BIND(ProcessData);
+
+
+    // scalar version
+    {
+      Label ScalarLoop;
+
+      Register byte0 = soff, byte1 = send, byte2 = doff, byte3 = isURL;
+      Register combined32Bits = x29; // t5
+
+      if (UseRVV) {
+        __ bnez(isMIME, ScalarLoop);
+        __ mv(t0, MaxVectorSize * 4);
+        __ bge(length, t0, ProcessVector);
+        __ BIND(ProcessScalar);
+        __ beqz(length, Exit);
+      }
+
+      __ BIND(ScalarLoop);
+      {
+        // encoded:   [byte0[5:0] : byte1[5:0] : byte2[5:0]] : byte3[5:0]] =>
+        // plain:     [byte0[5:0]+byte1[5:4] : byte1[3:0]+byte2[5:2] : byte2[1:0]+byte3[5:0]]
+
+        // load 4 bytes encoded src data
+        __ lbu(byte0, Address(src, 0));
+        __ lbu(byte1, Address(src, 1));
+        __ lbu(byte2, Address(src, 2));
+        __ lbu(byte3, Address(src, 3));
+        __ addi(src, src, 4);
+
+        // get codec index and decode (ie. load from codec by index)
+        __ add(byte0, codec, byte0);
+        __ add(byte1, codec, byte1);
+        __ lb(byte0, Address(byte0, 0));
+        __ lb(byte1, Address(byte1, 0));
+        __ add(byte2, codec, byte2);
+        __ add(byte3, codec, byte3);
+        __ lb(byte2, Address(byte2, 0));
+        __ lb(byte3, Address(byte3, 0));
+        __ slliw(byte0, byte0, 18);
+        __ slliw(byte1, byte1, 12);
+        __ orr(byte0, byte0, byte1);
+        __ orr(byte0, byte0, byte3);
+        __ slliw(byte2, byte2, 6);
```

Comment:
Is this correct to shift left and modify `byte2` before its original value is OR-ed into `combined32Bits`?

Reply:
Sorry for being not clear on this. The java code snippet of decodeBlock:
```
 795                 int b1 = base64[src[sp++] & 0xff];
 796                 int b2 = base64[src[sp++] & 0xff];
 797                 int b3 = base64[src[sp++] & 0xff];
 798                 int b4 = base64[src[sp++] & 0xff];
 799                 if ((b1 | b2 | b3 | b4) < 0) {    // non base64 byte
 800                     return new_dp - dp;
 801                 }
 802                 int bits0 = b1 << 18 | b2 << 12 | b3 << 6 | b4;
```
L799 simply OR-ed all the initial values of `b1`-`b4` and compare the result with zero. I think this should be reflected on the value of `combined32Bits` when it is used to do following error check. Correspondingly, It should be the OR-ed result of the initial loaded values in `byte0` - `byte3`.
```
      // error check
      __ bltz(combined32Bits, Exit);
```

Anything I missed?

Reply:
`(b1 | b2 | b3 | b4) < 0`, this java code is to tell if any of b1-b4 is < 0.
On the other side, `bltz(combined32Bits, Exit)` is doing the similar thing (same effect, but different way), as when loading byte1-4, `lb` will sign-extend it, if any of byte1-4 < 0, then the top bit will be 1, and after shift left, top bit will still be 1, so as a result, we can use combined32Bits to tell if any of byte1-4 < 0, and at the same time, we can use combined32Bits to decode the final data (from 4 bytes to 3 bytes).
Hope this answers your question?

Reply:
Yeah. Interesting. Could you please add a small code comment for this? Thanks.

Reply:
Sure, added. Thanks!